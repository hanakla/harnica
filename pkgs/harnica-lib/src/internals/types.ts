/** キー名の文字列表現 (e.g. 'C', 'D#', 'Fb') */
export type KeyString = string & { [keyStringType]: never };
declare const keyStringType: unique symbol;

const keyStringRegex = /^[A-G][#b]?$/;
export function assertKeyString(str: string): asserts str is KeyString {
  if (!keyStringRegex.test(str)) {
    throw new Error(`Invalid key string: ${str}`);
  }
}

export type BeatClock = [bars: number, beats: number, sixteeth: number];

export type NoteQuality =
  | [type: "quality", argument: "M" | "m"]
  | [type: "quality", argument: "M" | "m", argument: "7"]
  | [
      type: "omit" | "dim" | "aug" | "sus" | "add" | "tension",
      argument: string,
    ];

export type NoteMatch = {
  start: number;
  length: number;
  string: string;
};

export type NoteMatchFragment = {
  note: NoteParseResult | null;
  data: {
    key: string | null;
    sig: { [keyValue: number]: { isDegree: boolean; offset: number } };
    noteIdx: number | null;
    bar: number;
    notesInThisBar: number;
    isDegree: boolean | null;
    groupDepth: number;
    /** number of `type: 'note'` notes in this group. this number available only "beginBrace" note */
    notesInThisGroup: number | null;
  };
  match: NoteMatch;
};

export namespace NoteParseResult {
  export type Note = {
    type: "note";
    keyValues: number[];
    /** keys likes 'C3' */
    keys: string[];
    octave: -1 | 0 | 1;
    isDegree: boolean;
    resolveKey: string | null;
    detail: {
      root: string;
      degreeRoot: string;
      qualities: NoteQuality[];
      /** Degree or Alphabet notation chord name */
      originalChordName: string;
      /** Alphabet notation chord name (resolved or not) */
      chordName: string;
      warns: Array<{ type: string; message: string }>;
    };
    match: NoteMatch;
  };

  export type Repeat = { type: "repeat"; match: NoteMatch };

  export type Rest = { type: "rest"; match: NoteMatch };

  export type KeyChange = {
    type: "keyChange";
    key: string;
    match: NoteMatch;
  };

  export type SignatureChange = {
    type: "signatureChange";
    sigs: Array<{
      isDegree: boolean;
      targetKey: number;
      offset: number;
    }>;
    match: NoteMatch;
  };

  export type VoidSpace = {
    type: "voidSpace";
    match: NoteMatch;
  };

  export type Error = {
    type: "error";
    error: globalThis.Error;
    match: NoteMatch;
  };

  export type BraceBegin = {
    type: "braceBegin";
    match: NoteMatch;
  };

  export type BraceEnd = {
    type: "braceEnd";
    match: NoteMatch;
  };
}

export type NoteParseResult =
  | NoteParseResult.Note
  | NoteParseResult.Repeat
  | NoteParseResult.Rest
  | NoteParseResult.KeyChange
  | NoteParseResult.SignatureChange
  | NoteParseResult.VoidSpace
  | NoteParseResult.BraceBegin
  | NoteParseResult.BraceEnd
  | NoteParseResult.Error;
