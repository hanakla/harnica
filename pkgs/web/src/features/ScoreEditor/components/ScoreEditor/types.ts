export type OnReplaceNoteHandler = (noteIndex: number, newNote: string) => void;
export type OnInsertNoteAfterHandler = (
  beforeNoteIndex: number,
  newNote: string,
) => void;

export type OnMoveCursorToNote = (noteIndex: number) => void;
