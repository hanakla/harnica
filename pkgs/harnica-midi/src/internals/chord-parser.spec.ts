import { parseNoteQuality, parseSingleNote } from "./chord-parser";
import { NoteParseResult } from "./types";

describe.skip("chord-parser", () => {
  describe("parseNoteQuality", () => {
    it.each(
      // prettier-ignore
      [
        ['M7sus4', [['quality', 'M'], ['tension', '7'], ['sus', '4']]],
        ['m', [['quality', 'm']]],
        ['7', [['tension', '7']]],
        ['+7+9', [['tension', '7'], ['tension', '9']]],
        ['(#11)', [['tension', '#11']]],
        ['b9', [['tension', 'b9']]],
        ['b13', [['tension', 'b13']]],
        ['mM7', [['quality', 'm'], ['tension', 'M7']]],
        ['omit3', [['omit', '3']]],
        ['aug', [["aug", ""]]],
        ['aug7', [["aug", "7"]]],
        ["dim", [["dim", ""]]],
        ["dim7", [["dim", "7"]]],
      ],
    )("%s parsed to %s", (qualityStr, qualities) => {
      expect(parseNoteQuality(qualityStr)).toMatchObject(qualities);
    });
  });

  describe("parse notes to valid keys test", () => {
    it.only.each([
      ["C", ["C3", "E3", "G3"], [0, 4, 7]],
      ["Cm", ["C3", "D#3", "G3"], [0, 3, 7]],
      ["C7", ["C3", "E3", "G3", "A#3"], [0, 4, 7, 10]],
      ["Cm7", ["C3", "D#3", "G3", "A#3"], [0, 3, 7, 10]],
      ["Cmaj7", ["C3", "E3", "G3", "B3"], [0, 4, 7, 11]],
      ["CmM7", ["C3", "D#3", "G3", "B3"], [0, 3, 7, 11]],
      ["Csus4", ["C3", "F3", "G3"], [0, 5, 7]],
      ["Csus2", ["C3", "D3", "G3"], [0, 2, 7]],
      ["Caug", ["C3", "E3", "G#3"], [0, 4, 8]],
      ["Cdim", ["C3", "D#3", "F#3"], [0, 3, 6]],
      // ["Cm7b5", ["C3", "D#3", "F#3", "A#3"], [0, 3, 6, 10]],
    ])("parse %s", (input, expectedKeys, expectedValues) => {
      const note = parseSingleNote(input, "C") as NoteParseResult.ChordNote;
      expect(note.keys).toMatchObject(expectedKeys);
      expect(note.keyValues).toMatchObject(expectedValues);
    });
  });

  // describe("parseSingleNote", () => {
  //   it("English notations", () => {
  //     expect(parseSingleNote("C", "C")).toMatchInlineSnapshot(`
  //       {
  //         "detail": {
  //           "chordName": "C",
  //           "originalChordName": "C",
  //           "qualities": [],
  //           "root": "C",
  //           "warns": [],
  //         },
  //         "isDegree": false,
  //         "keyValues": [
  //           0,
  //           4,
  //           7,
  //         ],
  //         "keys": [
  //           "C3",
  //           "E3",
  //           "G3",
  //         ],
  //         "match": {
  //           "length": 1,
  //           "start": 0,
  //           "string": "C",
  //         },
  //         "octave": 0,
  //         "resolveKey": null,
  //         "type": "note",
  //       }
  //     `);

  //     expect(parseSingleNote("C-")).toMatchInlineSnapshot(`
  //       {
  //         "detail": {
  //           "chordName": "B",
  //           "originalChordName": "C-",
  //           "qualities": [],
  //           "root": "B",
  //           "warns": [],
  //         },
  //         "isDegree": false,
  //         "keyValues": [
  //           -1,
  //           3,
  //           6,
  //         ],
  //         "keys": [
  //           "B2",
  //           "D#3",
  //           "F#3",
  //         ],
  //         "match": {
  //           "length": 2,
  //           "start": 0,
  //           "string": "C-",
  //         },
  //         "octave": 0,
  //         "resolveKey": null,
  //         "type": "note",
  //       }
  //     `);
  //   });

  //   it.each(
  //     // prettier-ignore
  //     [
  //       ['IIsus4/V', { type: 'note', keyValues: [-5, 2, 7, 9], keys: ['G2', 'D3', 'G3', 'A3'], detail: { originalChordName: 'IIsus4/V' } }],
  //       ['III#M7(11)', { type: 'note', keyValues: [5, 9, 12, 15, 22], keys: ['F3', 'A3', 'C4', 'D#4', 'A#4'], detail: { originalChordName: 'III#M7(11)' } }],
  //       ['V#dim7/VII', { type: 'note', keyValues: [-1, 8, 11, 14, 17], keys: ['B2', 'G#3', 'B3', 'D4', 'F4'], detail: { originalChordName: 'V#dim7/VII' } }],
  //       // Spaced
  //       [' V#dim7/VII ', { type: 'note', keyValues: [-1, 8, 11, 14, 17], keys: ['B2', 'G#3', 'B3', 'D4', 'F4'], detail: { originalChordName: 'V#dim7/VII' } }]
  //     ]
  //   )("work with degree notation", (input, expected) => {
  //     expect(parseSingleNote(input)).toMatchObject(expected);
  //   });

  //   it("detect key change", () => {
  //     expect(parseSingleNote("Key=C#", "C")).toEqual({
  //       type: "keyChange",
  //       key: "C#",
  //       match: {
  //         start: 0,
  //         length: 6,
  //         string: "Key=C#",
  //       },
  //     });
  //   });

  //   it("omit", () => {
  //     const result1 = parseSingleNote("C7omit3", "C") as NoteParseResult.Note;
  //     expect(result1.keys).toEqual(["C3", "G3", "Bb3"]);

  //     const result2 = parseSingleNote("C7-3", "C") as NoteParseResult.Note;
  //     expect(result2.keys).toEqual(["C3", "G3", "A#3"]);
  //   });

  //   it("negative octave", () => {
  //     const result = parseSingleNote("-A", "C", 2) as NoteParseResult.Note;
  //     expect(result.keys).toEqual(["A1", "C#2", "E2"]);
  //   });

  //   it("positive octave", () => {
  //     const result = parseSingleNote("+A", "C", 2) as NoteParseResult.Note;
  //     expect(result.keys).toEqual(["A3", "C#4", "E4"]);
  //   });

  //   it("test", () => {
  //     // const matcher = /^([IV]+?[#b]?)([^/]*?)(?:\/([IV]+))?(omit\d+)?$/g;
  //     // console.log(matcher.exec("VII#maj7add9sus4/VII"));
  //   });
  // });

  // describe("parseChordProgression", () => {
  //   // it("test space", () => {
  //   //   console.log(
  //   //     parseChordProgression(
  //   //       `
  //   //     | Bm7 C#m7 | Em7 F#m7 | Em7 | F#m7 Bm7(11) |
  //   //     | Em7 F#m7 | Am7(9) F#sus4/G# | GM7(#11) CM7(#11) | FM7(#11)F#m/B |
  //   //   `,
  //   //       "C",
  //   //       4
  //   //     )
  //   //   );
  //   // });

  //   it("should parse chord progression", () => {
  //     const prog = `A#m7 _ # | AonG`;

  //     expect(parseChordProgression(prog, "C", 4)).toMatchSnapshot();
  //   });

  //   it("should parse with key change", () => {
  //     expect(parseChordProgression("C | Key=D I", "C", 4)).toMatchSnapshot();
  //   });

  //   it("should parse with Signature", () => {
  //     expect(
  //       parseChordProgression("Sig=Ab,Bb I | II", "C", 4)
  //     ).toMatchSnapshot();
  //   });

  //   it.skip("should parse with group", () => {
  //     const actual = parseChordProgression("[ C B ]  |", "C", 4);
  //     // console.log(actual);
  //     expect(pick(actual[0], "noteIdx", "key", "duration")).toMatchObject({
  //       noteIdx: 0,
  //       key: "C",
  //       duration: {
  //         beatBlock: [0, 2, 0],
  //       },
  //     });

  //     expect(actual[0]).toMatchObject({
  //       noteIdx: 1,
  //       key: "B",
  //       duration: {
  //         beatBlock: [0, 2, 0],
  //       },
  //     });
  //   });
  // });

  // describe("matchChordProgression", () => {
  //   it("should work", () => {
  //     const prog = matchChordProgression(
  //       " C#m7  | Am7 | G _ _ # | Em/B | Key=C"
  //     );

  //     expect(prog).toMatchSnapshot();
  //   });
  // });

  // describe("normalizeBeatClock", () => {
  //   it("work?", () => {
  //     expect(normalizeBeatClock([0, 4, 0], 4)).toMatchObject([1, 0, 0]);
  //     expect(normalizeBeatClock([0, 0, 4], 4)).toMatchObject([0, 1, 0]);
  //     expect(normalizeBeatClock([0, 0.5, 0], 4)).toMatchObject([0, 0, 2]);
  //   });
  // });

  // describe("sougo", () => {
  //   it("C#m7", () => {
  //     const note = parseSingleNote("C#m7") as NoteParseResult.Note;
  //     expect(getChordDetailFromKeyValues(note.keyValues).chordName).toBe(
  //       "C#m7"
  //     );
  //   });
  // });

  it.only.each([
    [`C#m7 | Am7 | G _ \n _ # | Em/B | Key=C`],
    [`Key=C\n\nC#m7 | Am7 | G _ \n _ # | Em/B \n`],
  ])("multi-line chord can be restore to original string", (prog) => {
    const restored = matchChordProgression(prog)
      .map((n) => n.match.string)
      .join("");

    expect(restored).toBe(prog);
  });
});

const pick = <T, K extends keyof T>(obj: T, ...keys: K[]): Pick<T, K> => {
  const ret = {} as Pick<T, K>;
  keys.forEach((key) => {
    ret[key] = obj[key];
  });
  return ret;
};
