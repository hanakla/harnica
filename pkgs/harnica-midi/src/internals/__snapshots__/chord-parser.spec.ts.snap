// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chord-parser > matchChordProgression > should work 1`] = `
[
  {
    "data": {
      "bar": 0,
      "groupDepth": 0,
      "isDegree": false,
      "key": null,
      "noteIdx": 0,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 4,
      "start": 1,
      "string": "C#m7",
    },
    "note": {
      "detail": {
        "chordName": "C#m7",
        "originalChordName": "C#m7",
        "qualities": [
          [
            "quality",
            "m",
          ],
          [
            "tension",
            "7",
          ],
        ],
        "root": "C#",
        "warns": [],
      },
      "isDegree": false,
      "keyValues": [
        1,
        4,
        8,
        11,
      ],
      "keys": [
        "C#3",
        "E3",
        "G#3",
        "B3",
      ],
      "match": {
        "length": 4,
        "start": 0,
        "string": "C#m7",
      },
      "octave": 0,
      "resolveKey": null,
      "type": "note",
    },
  },
  {
    "data": {
      "bar": 0,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": null,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 7,
      "string": "|",
    },
    "note": null,
  },
  {
    "data": {
      "bar": 1,
      "groupDepth": 0,
      "isDegree": false,
      "key": null,
      "noteIdx": 1,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 3,
      "start": 9,
      "string": "Am7",
    },
    "note": {
      "detail": {
        "chordName": "Am7",
        "originalChordName": "Am7",
        "qualities": [
          [
            "quality",
            "m",
          ],
          [
            "tension",
            "7",
          ],
        ],
        "root": "A",
        "warns": [],
      },
      "isDegree": false,
      "keyValues": [
        9,
        12,
        16,
        19,
      ],
      "keys": [
        "A3",
        "C4",
        "E4",
        "G4",
      ],
      "match": {
        "length": 3,
        "start": 0,
        "string": "Am7",
      },
      "octave": 0,
      "resolveKey": null,
      "type": "note",
    },
  },
  {
    "data": {
      "bar": 1,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": null,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 13,
      "string": "|",
    },
    "note": null,
  },
  {
    "data": {
      "bar": 2,
      "groupDepth": 0,
      "isDegree": false,
      "key": null,
      "noteIdx": 2,
      "notesInThisBar": 4,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 15,
      "string": "G",
    },
    "note": {
      "detail": {
        "chordName": "G",
        "originalChordName": "G",
        "qualities": [],
        "root": "G",
        "warns": [],
      },
      "isDegree": false,
      "keyValues": [
        7,
        11,
        14,
      ],
      "keys": [
        "G3",
        "B3",
        "D4",
      ],
      "match": {
        "length": 1,
        "start": 0,
        "string": "G",
      },
      "octave": 0,
      "resolveKey": null,
      "type": "note",
    },
  },
  {
    "data": {
      "bar": 2,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": 3,
      "notesInThisBar": 4,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 17,
      "string": "_",
    },
    "note": {
      "match": {
        "length": 1,
        "start": 0,
        "string": "_",
      },
      "type": "repeat",
    },
  },
  {
    "data": {
      "bar": 2,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": 4,
      "notesInThisBar": 4,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 19,
      "string": "_",
    },
    "note": {
      "match": {
        "length": 1,
        "start": 0,
        "string": "_",
      },
      "type": "repeat",
    },
  },
  {
    "data": {
      "bar": 2,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": 5,
      "notesInThisBar": 4,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 21,
      "string": "#",
    },
    "note": {
      "match": {
        "length": 1,
        "start": 0,
        "string": "#",
      },
      "type": "rest",
    },
  },
  {
    "data": {
      "bar": 2,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": null,
      "notesInThisBar": 4,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 23,
      "string": "|",
    },
    "note": null,
  },
  {
    "data": {
      "bar": 3,
      "groupDepth": 0,
      "isDegree": false,
      "key": null,
      "noteIdx": 6,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 4,
      "start": 25,
      "string": "Em/B",
    },
    "note": {
      "detail": {
        "chordName": "Em/B",
        "originalChordName": "Em/B",
        "qualities": [
          [
            "quality",
            "m",
          ],
        ],
        "root": "E",
        "warns": [],
      },
      "isDegree": false,
      "keyValues": [
        -1,
        4,
        7,
        11,
      ],
      "keys": [
        "B2",
        "E3",
        "G3",
        "B3",
      ],
      "match": {
        "length": 4,
        "start": 0,
        "string": "Em/B",
      },
      "octave": 0,
      "resolveKey": null,
      "type": "note",
    },
  },
  {
    "data": {
      "bar": 3,
      "groupDepth": 0,
      "isDegree": null,
      "key": null,
      "noteIdx": null,
      "notesInThisBar": 1,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 1,
      "start": 30,
      "string": "|",
    },
    "note": null,
  },
  {
    "data": {
      "bar": 4,
      "groupDepth": 0,
      "isDegree": null,
      "key": "C",
      "noteIdx": null,
      "notesInThisBar": 0,
      "notesInThisGroup": null,
      "sig": {},
    },
    "match": {
      "length": 5,
      "start": 32,
      "string": "Key=C",
    },
    "note": {
      "key": "C",
      "match": {
        "length": 5,
        "start": 0,
        "string": "Key=C",
      },
      "type": "keyChange",
    },
  },
]
`;

exports[`chord-parser > parseChordProgression > should parse chord progression 1`] = `
[
  {
    "chord": "A#m7",
    "duration": {
      "beatClock": [
        0,
        1,
        1,
      ],
      "beatClockStr": "0:1:1",
      "tick": "T171",
    },
    "key": "C",
    "keyValues": [
      10,
      13,
      17,
      20,
    ],
    "keys": [
      "A#3",
      "C#4",
      "F4",
      "G#4",
    ],
    "match": {
      "data": {
        "bar": 0,
        "groupDepth": 0,
        "isDegree": false,
        "key": null,
        "noteIdx": 0,
        "notesInThisBar": 3,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 4,
        "start": 0,
        "string": "A#m7",
      },
      "note": {
        "detail": {
          "chordName": "A#m7",
          "originalChordName": "A#m7",
          "qualities": [
            [
              "quality",
              "m",
            ],
            [
              "tension",
              "7",
            ],
          ],
          "root": "A#",
          "warns": [],
        },
        "isDegree": false,
        "keyValues": [
          10,
          13,
          17,
          20,
        ],
        "keys": [
          "A#3",
          "C#4",
          "F4",
          "G#4",
        ],
        "match": {
          "length": 4,
          "start": 0,
          "string": "A#m7",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 0,
    "sig": {},
    "startAt": {
      "beatClock": [
        0,
        0,
        0,
      ],
      "beatClockStr": "0:0:0",
    },
  },
  {
    "chord": "A#m7",
    "duration": {
      "beatClock": [
        0,
        1,
        1,
      ],
      "beatClockStr": "0:1:1",
      "tick": "T171",
    },
    "key": "C",
    "keyValues": [
      10,
      13,
      17,
      20,
    ],
    "keys": [
      "A#3",
      "C#4",
      "F4",
      "G#4",
    ],
    "match": {
      "data": {
        "bar": 0,
        "groupDepth": 0,
        "isDegree": null,
        "key": null,
        "noteIdx": 1,
        "notesInThisBar": 3,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 1,
        "start": 5,
        "string": "_",
      },
      "note": {
        "match": {
          "length": 1,
          "start": 0,
          "string": "_",
        },
        "type": "repeat",
      },
    },
    "noteIdx": 1,
    "sig": {},
    "startAt": {
      "beatClock": [
        0,
        1,
        1,
      ],
      "beatClockStr": "0:1:1",
    },
  },
  {
    "chord": null,
    "duration": {
      "beatClock": [
        0,
        1,
        1,
      ],
      "beatClockStr": "0:1:1",
      "tick": "T171",
    },
    "key": "C",
    "keyValues": [],
    "keys": [],
    "match": {
      "data": {
        "bar": 0,
        "groupDepth": 0,
        "isDegree": null,
        "key": null,
        "noteIdx": 2,
        "notesInThisBar": 3,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 1,
        "start": 7,
        "string": "#",
      },
      "note": {
        "match": {
          "length": 1,
          "start": 0,
          "string": "#",
        },
        "type": "rest",
      },
    },
    "noteIdx": 2,
    "sig": {},
    "startAt": {
      "beatClock": [
        0,
        2,
        2,
      ],
      "beatClockStr": "0:2:2",
    },
  },
  {
    "chord": "A",
    "duration": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
      "tick": "T512",
    },
    "key": "C",
    "keyValues": [
      9,
      13,
      16,
    ],
    "keys": [
      "A3",
      "C#4",
      "E4",
    ],
    "match": {
      "data": {
        "bar": 1,
        "groupDepth": 0,
        "isDegree": false,
        "key": null,
        "noteIdx": 3,
        "notesInThisBar": 1,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 4,
        "start": 11,
        "string": "AonG",
      },
      "note": {
        "detail": {
          "chordName": "A",
          "originalChordName": "AonG",
          "qualities": [],
          "root": "A",
          "warns": [],
        },
        "isDegree": false,
        "keyValues": [
          9,
          13,
          16,
        ],
        "keys": [
          "A3",
          "C#4",
          "E4",
        ],
        "match": {
          "length": 4,
          "start": 0,
          "string": "AonG",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 3,
    "sig": {},
    "startAt": {
      "beatClock": [
        0,
        3,
        3,
      ],
      "beatClockStr": "0:3:3",
    },
  },
]
`;

exports[`chord-parser > parseChordProgression > should parse with Signature 1`] = `
[
  {
    "chord": "C",
    "duration": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
      "tick": "T512",
    },
    "key": "C",
    "keyValues": [
      0,
      4,
      7,
    ],
    "keys": [
      "C3",
      "E3",
      "G3",
    ],
    "match": {
      "data": {
        "bar": 0,
        "groupDepth": 0,
        "isDegree": true,
        "key": null,
        "noteIdx": 0,
        "notesInThisBar": 1,
        "notesInThisGroup": null,
        "sig": {
          "11": {
            "isDegree": false,
            "offset": -1,
          },
          "9": {
            "isDegree": false,
            "offset": -1,
          },
        },
      },
      "match": {
        "length": 1,
        "start": 10,
        "string": "I",
      },
      "note": {
        "detail": {
          "chordName": "C",
          "originalChordName": "I",
          "qualities": [],
          "root": "C",
          "warns": [],
        },
        "isDegree": true,
        "keyValues": [
          0,
          4,
          7,
        ],
        "keys": [
          "C3",
          "E3",
          "G3",
        ],
        "match": {
          "length": 1,
          "start": 0,
          "string": "I",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 0,
    "sig": [
      {
        "isDegree": false,
        "offset": -1,
        "targetKey": 9,
      },
      {
        "isDegree": false,
        "offset": -1,
        "targetKey": 11,
      },
    ],
    "startAt": {
      "beatClock": [
        0,
        0,
        0,
      ],
      "beatClockStr": "0:0:0",
    },
  },
  {
    "chord": "D",
    "duration": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
      "tick": "T512",
    },
    "key": "C",
    "keyValues": [
      2,
      6,
      9,
    ],
    "keys": [
      "D3",
      "F#3",
      "A3",
    ],
    "match": {
      "data": {
        "bar": 1,
        "groupDepth": 0,
        "isDegree": true,
        "key": null,
        "noteIdx": 1,
        "notesInThisBar": 1,
        "notesInThisGroup": null,
        "sig": {
          "11": {
            "isDegree": false,
            "offset": -1,
          },
          "9": {
            "isDegree": false,
            "offset": -1,
          },
        },
      },
      "match": {
        "length": 2,
        "start": 14,
        "string": "II",
      },
      "note": {
        "detail": {
          "chordName": "D",
          "originalChordName": "II",
          "qualities": [],
          "root": "D",
          "warns": [],
        },
        "isDegree": true,
        "keyValues": [
          2,
          6,
          9,
        ],
        "keys": [
          "D3",
          "F#3",
          "A3",
        ],
        "match": {
          "length": 2,
          "start": 0,
          "string": "II",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 1,
    "sig": [
      {
        "isDegree": false,
        "offset": -1,
        "targetKey": 9,
      },
      {
        "isDegree": false,
        "offset": -1,
        "targetKey": 11,
      },
    ],
    "startAt": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
    },
  },
]
`;

exports[`chord-parser > parseChordProgression > should parse with key change 1`] = `
[
  {
    "chord": "C",
    "duration": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
      "tick": "T512",
    },
    "key": "C",
    "keyValues": [
      0,
      4,
      7,
    ],
    "keys": [
      "C3",
      "E3",
      "G3",
    ],
    "match": {
      "data": {
        "bar": 0,
        "groupDepth": 0,
        "isDegree": false,
        "key": null,
        "noteIdx": 0,
        "notesInThisBar": 1,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 1,
        "start": 0,
        "string": "C",
      },
      "note": {
        "detail": {
          "chordName": "C",
          "originalChordName": "C",
          "qualities": [],
          "root": "C",
          "warns": [],
        },
        "isDegree": false,
        "keyValues": [
          0,
          4,
          7,
        ],
        "keys": [
          "C3",
          "E3",
          "G3",
        ],
        "match": {
          "length": 1,
          "start": 0,
          "string": "C",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 0,
    "sig": {},
    "startAt": {
      "beatClock": [
        0,
        0,
        0,
      ],
      "beatClockStr": "0:0:0",
    },
  },
  {
    "chord": "C",
    "duration": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
      "tick": "T512",
    },
    "key": "D",
    "keyValues": [
      2,
      6,
      9,
    ],
    "keys": [
      "D3",
      "F#3",
      "A3",
    ],
    "match": {
      "data": {
        "bar": 1,
        "groupDepth": 0,
        "isDegree": true,
        "key": "D",
        "noteIdx": 1,
        "notesInThisBar": 1,
        "notesInThisGroup": null,
        "sig": {},
      },
      "match": {
        "length": 1,
        "start": 10,
        "string": "I",
      },
      "note": {
        "detail": {
          "chordName": "C",
          "originalChordName": "I",
          "qualities": [],
          "root": "C",
          "warns": [],
        },
        "isDegree": true,
        "keyValues": [
          0,
          4,
          7,
        ],
        "keys": [
          "C3",
          "E3",
          "G3",
        ],
        "match": {
          "length": 1,
          "start": 0,
          "string": "I",
        },
        "octave": 0,
        "resolveKey": null,
        "type": "note",
      },
    },
    "noteIdx": 1,
    "sig": {},
    "startAt": {
      "beatClock": [
        1,
        0,
        0,
      ],
      "beatClockStr": "1:0:0",
    },
  },
]
`;
